polkadot/primitives/src/vstaging/mod.rs

```diff
@@ -465,11 +505,72 @@ impl CandidateCommitments {
 			{
 				UMPSignal::SelectCore(core_index_selector, cq_offset) =>
 					Ok(Some((core_index_selector, cq_offset))),
+				_ => Err(CommittedCandidateReceiptError::UmpSignalDecode),
 			}
 		} else {
 			Ok(None)
 		}
 	}
+
+	/// Returns the ump signals of this candidate, if any, or an error if they violate the expected
+	/// format.
+	pub fn ump_signals(
+		&self,
+		allow_approved_peer_signal: bool,
+	) -> Result<CandidateUMPSignals, CommittedCandidateReceiptError> {
+		let mut res = CandidateUMPSignals { select_core: None, approved_peer: None };
+
+		if !allow_approved_peer_signal {
+			res.select_core = self.core_selector_before_approved_peer_allowed()?;
+
+			return Ok(res)
+		}
+
+		let mut signals_iter =
+			self.upward_messages.iter().skip_while(|message| *message != &UMP_SEPARATOR);
+
+		if signals_iter.next().is_none() {
+			// No UMP separator
+			return Ok(res)
+		}
+
+		// Process first signal
+		let Some(first_signal) = signals_iter.next() else { return Ok(res) };
+		match UMPSignal::decode(&mut first_signal.as_slice())
+			.map_err(|_| CommittedCandidateReceiptError::UmpSignalDecode)?
+		{
+			UMPSignal::ApprovedPeer(approved_peer_id) => {
+				res.approved_peer = Some(approved_peer_id);
+			},
+			UMPSignal::SelectCore(core_selector, cq_offset) => {
+				res.select_core = Some((core_selector, cq_offset));
+			},
+		};
+
+		// Process second signal
+		let Some(second_signal) = signals_iter.next() else { return Ok(res) };
+		match UMPSignal::decode(&mut second_signal.as_slice())
+			.map_err(|_| CommittedCandidateReceiptError::UmpSignalDecode)?
+		{
+			UMPSignal::ApprovedPeer(approved_peer_id) if res.approved_peer.is_none() => {
+				res.approved_peer = Some(approved_peer_id);
+			},
+			UMPSignal::SelectCore(core_selector, cq_offset) if res.select_core.is_none() => {
+				res.select_core = Some((core_selector, cq_offset));
+			},
+			_ => {
+				// This means that we got duplicate UMP signals.
```

Comment:
How does this look if we add another signal ? Can this be re-written to deduplicate ?

Reply:
Wrote this in a more generic way